src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java

```diff
@@ -144,4 +167,99 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
+            super(publisher.getMaxSchedule(), publisher.getDelay(), publisher.isRerunMatrixPart());
+            this.regexpForRerun = publisher.getRegexpForRerun();
+            this.rerunIfUnstable = publisher.isRerunIfUnstable();
+            this.checkRegexp = publisher.isCheckRegexp();
+        }
+        
+        @Nullable
+        public String getRegexpForRerun() {
+            return regexpForRerun;
+        }
+        
+        public boolean isRerunIfUnstable() {
+            return rerunIfUnstable;
+        }
+        
+        public boolean isCheckRegexp() {
+            return checkRegexp;
+        }
+        
+        @Override
+        public boolean shouldSchedule(@Nonnull Run<?, ?> run, @Nonnull TaskListener listener, int retryCount) {
+            if ((run.getResult() == Result.SUCCESS) || (run.getResult() == Result.ABORTED)) {
+                return false;
+            }
+            
+            // If we're not set to rerun if unstable, and the build's unstable, return true.
+            if ((!isRerunIfUnstable()) && (run.getResult() == Result.UNSTABLE)) {
+                return false;
+            }
+            
+            // If we're supposed to check for a regular expression in the build output before
+            // scheduling a new build, do so.
+            if (isCheckRegexp()) {
+                LOGGER.log(Level.FINEST, "Got checkRegexp == true");
+                
+                String regexpForRerun = getRegexpForRerun();
+                if ((regexpForRerun != null) && (!regexpForRerun.equals(""))) {
+                    LOGGER.log(Level.FINEST, "regexpForRerun - {0}", regexpForRerun);
+                    
+                    try {
+                        // If parseLog returns false, we didn't find the regular expression,
+                        // so return true.
+                        if (!parseLog(run.getLogFile(), regexpForRerun)) {
+                            LOGGER.log(Level.FINEST, "regexp not in logfile");
+                            return false;
+                        }
+                    } catch (IOException e) {
+                        e.printStackTrace(listener
+                                          .error("error while parsing logs for naginator - forcing rebuild."));
+                    }
+                }
+            }
+            
+            return super.shouldSchedule(run, listener, retryCount);
+        }
+
+        @Override
+        public boolean shouldScheduleForMatrixRun(@Nonnull MatrixRun run, @Nonnull TaskListener listener) {
+            if ((run.getResult() == Result.SUCCESS) || (run.getResult() == Result.ABORTED)) {
+                return false;
+            }
+            if ((!isRerunIfUnstable()) && (run.getResult() == Result.UNSTABLE)) {
+                return false;
+            }
+            return true;
+        }
+        
+        private boolean parseLog(File logFile, @Nonnull String regexp) throws IOException {
```

Comment:
And this annotations will be used only on pom jenkins version, because they are required only for build time. So you should add them and remove after base upgrade.